File: proposed-projects/round-trip-experiment/reverse/20260909T134433Z_3f62598c/20260910T080424Z_8686d09e/ported_implementation/src/rulesBuilder/symbolIds.ts
export class SymbolIds {
  // we don't need to delete, just preserve relationships
  private map = new Map<string, number>();
  private pos = new Map<string, number>();
  private reverseMap = new Map<number, string>();

  get size(): number {
    return this.map.size;
  }

  keys(): IterableIterator<string> {
    return this.map.keys();
  }

  has(key: string): boolean {
    return this.map.has(key);
  }

  get(key: string): number {
    if (!this.map.has(key)) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return this.map.get(key) as number;
  }

  reverseGet(key: number): string {
    if (!this.reverseMap.has(key)) {
      throw new Error(`SymbolIds does not contain value: ${key}`);
    }
    return this.reverseMap.get(key) as string;
  }

  getPos(key: string): number {
    if (!this.pos.has(key)) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return this.pos.get(key) as number;
  }

  set(key: string, value: number, pos: number): void {
    this.map.set(key, value);
    this.pos.set(key, pos);
    this.reverseMap.set(value, key);
  }

  [Symbol.iterator](): IterableIterator<[string, number]> {
    return [...this.map.entries()][Symbol.iterator]();
  }
}

File: packages/gbnf-experiment/data/20260909T040730Z_0947f229/ported_implementation/gbnf/rules_builder/symbol_ids.py
"""Port of ``src/rules-builder/symbol-ids.ts``."""

from __future__ import annotations

from typing import Dict, Iterator, Tuple


class SymbolIds:
    # we don't need to delete, just preserve relationships
    def __init__(self) -> None:
        self._map: Dict[str, int] = {}
        self._pos: Dict[str, int] = {}
        self._reverse_map: Dict[int, str] = {}

    @property
    def size(self) -> int:
        return len(self._map)

    def keys(self) -> Iterator[str]:
        return iter(list(self._map.keys()))

    def has(self, key: str) -> bool:
        return key in self._map

    def get(self, key: str) -> int:
        val = self._map.get(key)
        if val is None:
            raise KeyError(f'SymbolIds does not contain key: {key}')
        return val

    def reverse_get(self, key: int) -> str:
        val = self._reverse_map.get(key)
        if val is None:
            raise KeyError(f'SymbolIds does not contain value: {key}')
        return val

    def get_pos(self, key: str) -> int:
        pos = self._pos.get(key)
        if pos is None:
            raise KeyError(f'SymbolIds does not contain key: {key}')
        return pos

    def set(self, key: str, value: int, pos: int) -> None:
        self._map[key] = value
        self._pos[key] = pos
        self._reverse_map[value] = key

    def __iter__(self) -> Iterator[Tuple[str, int]]:
        return iter(list(self._map.items()))

    def __len__(self) -> int:
        return len(self._map)
